### PR TITLE
[IMP] ir_actions, safe_eval: better traceback/debug

### PR DESCRIPTION
**Pull request by Elkasitu**
Problem 1:

It was completely impossible to debug a server
action using pdb (or any other python debugger for that matter),
in order for pdb to work properly it must read the
file that it's supposed to debug from disk, since the code for server
actions is executed from memory with safe_eval, pdb can't display
any relevant information of the code being debugged.

Problem 2:

Tracebacks for safe_eval'd code were pretty useless, since all they
would do would be display the entire code that failed execution in raw
format (single-line), this is obviously very hard to debug and offers
absolutely no insight into which server action broke or where and why
did it break.

Solution 1:

Before compiling code that is supposed to be safe_eval'd, we create a
temporary file, this temporary file can have a suffix containing
information of the location of the code (e.g. '_server_action_<id>') and
where it should be stored (filestore location recommended), this file
will contain the server action code (or any code being safe_eval'd) so
that it can be debugged. This information must be passed explicitly from
the calling code to safe_eval, through a context argument which is a
dictionary with `fname` and `filestore` as keys, the former being an
indication of the calling code (e.g. '_server_action_<id>') and the
latter being the current db's filestore, so as to avoid file conflicts.
After the code is executed, the file is destroyed from the filesystem.

Solution 2:

Thanks to solution 1, the traceback will print out the corresponding
error message and the name of the file which ideally should contain
information about the calling code (e.g. 'File
~/.local/Odoo/filestore/my_db/tmphHfj53j_server_action_55') as well as
the line inside the code that triggered the traceback, effectively
indetifying where the code broke.

This commit also merges the `record` and `records` variables available
in server actions into a single variable `self`, so that server action
development is more consistent with native Odoo development, however
both record and records are still kept for backwards compatibility.

~~This commit also adds the possibility to call pdb.set_trace() from
within a server action by simply calling `pdb()`, to be reviewed later
to support multiple debuggers (py37's breakpoint() comes to mind).~~

Task ID: 54014

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_actions.py [hunk 1]
```diff
@@ -498,25 +503,29 @@ def log(message, level="info"):
...
             'env': self.env,
             'model': model,
             # Exceptions
             'Warning': odoo.exceptions.Warning,
             # record
-            'record': record,
-            'records': records,
+            'record': records, # backwards compatibility
+            'records': records, # backwards compatibility
+            'self': records,
```
Comment (tfr-odoo) [inline odoo/addons/base/models/ir_actions.py]: Not sure that putting PDB for real in the eval context is a good idea, we just add it locally. Anyway we cannot debug from a saas server. It doesn't make much sense. Maybe just keep the line in comment.

Comment (Elkasitu) [reply odoo/addons/base/models/ir_actions.py]: you're right, updated :^)

**Comment by JKE-be** [comment]
**Rename records** to self and merge both is probably just useless!
'# backwards compatibility', meaning that you plan to change all server action in code and all custom ? or just break it by renaming for pleasure in next version once removed?
" so that server action development is more consistent with native Odoo development" => in an action server, if I have the self, I'm waiting to have the action server... so really more Odoo dev oriented?

**Pdb** was just the worst idea from this PR for a production server, if you need it, check at least that you are in dev mode first!

**NamedTemporaryFile** just to raise the current server action that is broken ???
Why not just give it as name like suggested by documentation:

> compile() Parameters
> - filename - file from which the code was read. **If it wasn't read from a file, you can give a name yourself**


```python
>>> eval(compile("3+0", "Your_buggy_actionserver", "eval"))
3
>>> eval(compile("3/0", "Your_buggy_actionserver", "eval"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "Your_buggy_actionserver", line 1, in <module>
ZeroDivisionError: integer division or modulo by zero
```


**raise Exception** instead of ValueError, the PR could explain the real purpose behind this change?
Probably side effect in most code that use safe eval like https://github.com/odoo/odoo/blob/11.0/addons/gamification/models/goal.py#L89

About the **|= for onchange**, that is maybe correct, or not...  why change it? and why in master and not in other branch ? what is the fix / imp reason to change that ?

**Comment by Elkasitu** [comment]
> Rename records to self and merge both is probably just useless!

Well this feature was requested by the PS Tech, whom use server actions regularly...

> '# backwards compatibility', meaning that you plan to change all server action in code and all custom ? or just break it by renaming for pleasure in next version once removed?

I don't see what the problem is, old SAs (our own and custom) should still work and new ones *should* use the new `self` syntax, there is no need to remove the backwards compatibility at any point as there is no cost to keeping `record` and `records` forever.

> " so that server action development is more consistent with native Odoo development" => in an action server, if I have the self, I'm waiting to have the action server... so really more Odoo dev oriented?

You can choose a model for server actions, `self` refers to that model, originally I had named it `records` originally but @rco-odoo suggested `self` instead and I agreed.

> Pdb was just the worst idea from this PR for a production server, if you need it, check at least that you are in dev mode first!

Indeed, that is why it was removed from the original PR, I updated the PR description to reflect that.

> NamedTemporaryFile just to raise the current server action that is broken ???

No, the feature request was to be able to debug SA code by putting a pdb near the part of the code that executes the `eval`, however, in python, it seems impossible to debug code that is in memory (your pdb shows nothing, not which line you're at, doesn't show the code, anything), therefore to be able to debug python code pdb **requires** a file, that is why the file is created, the name is just to kill two birds with one stone.

> raise Exception instead of ValueError, the PR could explain the real purpose behind this change?
Probably side effect in most code that use safe eval like https://github.com/odoo/odoo/blob/11.0/addons/gamification/models/goal.py#L89

This was a hack to be able to display a pseudo-traceback when a SA raised an exception, which is no longer needed since with the temp file a proper traceback is obtained. Wouldn't it raise a `SyntaxError` and be caught anyway? I did a quick grep and found out that most safe_evals inside a try/except simply check for `Exception`, the gamification one is the only one I could find that explicitly checks for `ValueError`, but thanks for bringing this to my attention, I'll take a closer look.

> About the |= for onchange, that is maybe correct, or not... why change it? and why in master and not in other branch ? what is the fix / imp reason to change that ?

I'm actually not sure why I did that, I might revert that part, thanks.

**Pull request closed without merging**